File: leetcode/java/0152. Maximum Product Subarray.java
package leetcode.java;

class Solution {
    public int maxProduct(int[] nums) {
        int currMin = 1;
        int currMax = 1;
        int maximum = -99;

        for (int num : nums) {
            int tempMax = currMax * num;
            currMax = Integer.max(Integer.max(currMin*num, tempMax), num);
            currMin = Integer.min(Integer.min(currMin*num, tempMax), num);

            maximum = Integer.max(maximum, currMax);
        }
        return maximum;
    }
}

File: leetcode/python/0152. Maximum Product Subarray.py
class Solution(object):
    def maxProduct(self, nums):
        curr_min = 1
        curr_max = 1
        global_max = -99999999

        for n in nums:
            temp_max = curr_max * n
            curr_max = max(n, temp_max, curr_min*n)
            curr_min = min(n, temp_max, curr_min*n)
            global_max = max(curr_max, global_max)
        return global_max